MobiusCore/Source/EffectHandlers/EffectHandler.swift

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+import Foundation
+
+public enum HandleEffect<Type> {
+    case handle(Type)
+    case ignore
+}
+
+final public class EffectHandler<Effect, Event> {
+    private let lock = NSRecursiveLock()
+    private let handleEffect: (Effect, @escaping Consumer<Event>) -> Void
+    private let canAcceptEffect: (Effect) -> Bool
+    private let disposeFn: () -> Void
+    private var consumer: Consumer<Event>?
+
+    /// Create a handler for effects which satisfy the `canAccept` parameter function.
+    ///
+    /// - Parameter canAccept: A function which indicates whether to handle an effect. If it returns `.handle(effect)` for a given `effect`, this
+    /// effect handler will handle said effect.
+    /// - Parameter handleEffect: Handle effects which satisfy `canAccept`.
+    /// - Parameter onDispose: Tear down any resources being used by this effect handler.
+    public init<AssociatedValueType>(
+        canAccept: @escaping (Effect) -> HandleEffect<AssociatedValueType>,
+        handleEffect: @escaping (AssociatedValueType, @escaping Consumer<Event>) -> Void,
+        onDispose disposable: @escaping () -> Void
```

Comment:
The biggest point-of-use difference between our implementations is that I bundled these three functions into a protocol. Apart from a general feeling that “protocols are more Swifty”, I found that this was quite nice to use in that it felt natural to group the `canAccept` and `handleEffect` parts of each effect into a struct or final class depending on statefulness. It also led to composition like this:

```swift
func myFeatureEffectHandler(thingManager: ThingManager) -> CompositeEffectHandler<MyLoopTypes> {
    return CompositeEffectHandler()
        .addHandler(MyFirstEffectHandler())
        .addHandler(MySecondEffectHandler(thingManager: thingManager))
}
```

How will features look with the closure-based design? I imagine developers will either end up bundling effect handlers in types anyway to manage state, or creating a hierarchy of factory functions with a bunch of closure captures that may make debugging and memory management harder.

Reply:
I want to avoid the protocol-first approach. Mainly due to how generics are handled. By using a protocol as our base abstraction we'd need to introduce a type-erased `AnyEffectHandler` variant which adds confusion and pollutes the APIs.

I would use this `EffectHandler<Effect, Event>` type as our base abstraction. That way we can write code which depends on the generic idea of an `EffectHandler<F,E>` instead of concrete implementations or type-erasing wrappers.

Using this as our base abstraction also allows users to build their own protocol(s) if they wish, as long as they can supply a function `MyEffectHandlerProtocol -> EffectHandler<..>`. 

This closure based approach still allows you to use classes or structs to hold onto the effect handler state. Your example could, for example, be written as:
```
func myFeatureEffectHandler(thingManager: ThingManager) -> CompositeEffectHandler<MyLoopTypes> {
    return CompositeEffectHandler()
        .addHandler(MyFirstEffectHandler.make())
        .addHandler(MySecondEffectHandler.make(thingManager: thingManager))
}
```

---

I also think there is a deeper issue here. An effect handler should really just be a way of tying a side-effecting construct into a Mobius loop:

For instance, say you have something which can download a song. This "thing" will take some input (which song to download), and produce some output (potentially the downloaded song, download progress updates, errors etc). All of these things exist outside Mobius, and this thing should therefore live outside of Mobius too.

This "thing" is not an effect handler in an of itself. However, it is likely something we want to provide to the loop. An EffectHandler is just an adapter layer which maps such a generic input-output "thing" into a loop. I think the API I am suggesting makes that clearer.

Reply:
Hmm. I’m seeing quite complex effect handlers in our codebase, but maybe that’s an antipattern. If we actively promote a model where one effect handler binds one effect to one collaborator, the closure-based approach might actually help.

I don’t particularly see the `AnyEffectHandler` issue as a strong argument. This is a well-known wart that Swift developers should get used to, and APIs can be designed so that client code never needs to use `AnyEffectHandler` directly.

In any case, this is more syntax than semantics and we could easily transition from one approach to the other without breaking compatibility, so I guess I’m OK with either for now.